### ISSUE TITLE
[FEATURE REQUEST]: Option to hide SVGs that were optimised by 0%

### ISSUE THREAD

**Issue by JackDomleoOcean** [svg/svgo#2050]
**Is your feature request related to a problem? Please describe.**

When dealing with a huge library of SVGs, it's often difficult to see exactly which icons were optimised in the CLI output and by how much since there are a lot that were optimised by 0%.

**Describe the solution you'd like**

A new CLI flag and/or config property to hide any SVGs that were optimised by 0% in the output, so only the SVGs that were altered are displayed in the CLI output.

**Describe alternatives you've considered**

There is a `--quiet` flag, but this hides everything. I'd still like to see the SVGs that were optimised.

**Additional context**

![feature-request-hide-0](https://github.com/user-attachments/assets/186d63b4-77a5-4851-a9cc-c68b01a680f0)

**Comment by viralcodex** [comment]
Hi @SethFalco, would love to work on this one if it is fine.
Thanks 😄

**Comment by viralcodex** [comment]
Hi @SethFalco,

The PR for this feature is open now, you may have a look 😄 

Thanks

**Comment by GreLI** [comment]
Personally, I think that optimizing SVG files in CI generally is a bad idea. You're wasting energy, cpu cycles and time to optimize what is already optimized or can be optimized only once. By hiding SVG that weren't optimized you only hiding the problem instead of solving it. This can be solved by caching or changing your workflow.

**Comment by viralcodex** [comment]
> Personally, I think that optimizing SVG files in CI generally is a bad idea. You're wasting energy, cpu cycles and time to optimize what is already optimized or can be optimized only once. By hiding SVG that weren't optimized you only hiding the problem instead of solving it. This can be solved by caching or changing your workflow.

I think people still use it in CLI, we can't say how and what they will use unfortunately, maybe they have a lot of SVGs in their folder and want to get them to get optimised in a go. It is a good accessibility option for users to have, not something that is really needed but is a good to have.

**Comment by SethFalco** [comment]
Admittedly, I'm not sure if this is something we should handle on our end either.

I'd like to hear a bit more context before we commit to maintaining another argument for our CLI.

My understanding is that this is for knowing which SVGs were affected when bulk processing. However, the solution to this is usually version control, like Git. I don't see conditional output in SVGO being a practical solution here.

And if one isn't using version control, I think it's a really bad idea to bulk optimize SVGs, unless they're being output to another directory or as part of a build process. (i.e. SVGR/webpack) However, in those cases a human isn't reviewing CLI output, so this obvious won't help with that.

I won't close this immediately, as I'm open to feedback! So I'd appreciate if anyone with more context on this use case could help me understand the benefit.

However, I do think this is better suited to being resolved with external tools. For example:

* Use version control to track changes.
* Or use grep, i.e. `svgo ./* | grep -P ' (?!0%)[\d.]+%' -B 2` which only shows files that become smaller.

I also find the request to be a bit oddly worded. At least as a user myself, I wouldn't want to only know which files changed size, I'd want to know which files were modified by SVGO. For example, if SVGO failed to reduce the total number of bytes, yet still made modifications (such as sorting attributes) I wouldn't want that to be hidden.

So (hypothetically) if we set out to implement this, I'd expect the implementation to check if the SVG contents has changed, not only if the size changed.

**Comment by viralcodex** [comment]
> Admittedly, I'm not sure if this is something we should handle on our end either.
> 
> I'd like to hear a bit more context before we commit to maintaining another argument for our CLI.
> 
> My understanding is that this is for knowing which SVGs were affected when bulk processing. However, the solution to this is usually version control, like Git. I don't see conditional output in SVGO being a practical solution here.
> 
> And if one isn't using version control, I think it's a really bad idea to bulk optimize SVGs, unless they're being output to another directory or as part of a build process. (i.e. SVGR/webpack) However, in those cases a human isn't reviewing CLI output, so this obvious won't help with that.
> 
> I won't close this immediately, as I'm open to feedback! So I'd appreciate if anyone with more context on this use case could help me understand the benefit.
> 
> However, I do think this is better suited to being resolved with external tools. For example:
> 
> * Use version control to track changes.
> * Or use grep, i.e. `svgo ./* | grep -P ' (?!0%)[\d.]+%' -B 2` which only shows files that become smaller.
> 
> I also find the request to be a bit oddly worded. At least as a user myself, I wouldn't want to only know which files changed size, I'd want to know which files were modified by SVGO. For example, if SVGO failed to reduce the total number of bytes, yet still made modifications (such as sorting attributes) I wouldn't want that to be hidden.
> 
> So (hypothetically) if we set out to implement this, I'd expect the implementation to check if the SVG contents has changed, not only if the size changed.

Hi @SethFalco,

After considering and looking at my own use cases where I think it can work for me, we can have a small flag in the case where I have a folder for large number of svgs and I want to have a workflow run to reduce them, some may get reduced, other's don't.
Especially useful for icons library in my opinion, we can easily hide the unchanged svgs and just focus on the changed ones.

Also, since we provide the option, users won't have to deal with platform specific tools/commands for this case, like grep doesn't work on Windows, and regex are generally complex to implement and use and they break easily under changing requirements.

So, instead of just checking the size,  if there is no change (expanding the scope of the feature) apart from the size in svg, we should have a native flag just to hide the unchanged svg in any way. 

This will also help (similar to my use case) in build processes for logs (I guess) when there are a lot of svgs to handle, this flag will help separating unchanged ones from optimised ones in logs. 

Lastly the code change is just 5-6 lines and nothing impactful to original codebase and it's functionalities.

I think it will be a useful addition albeit a small one, like linters also give the option in CLI to hide unchanged files.

Let me know what you think and we can discuss more on it 😄

**Comment by GreLI** [comment]
So many questions. 

Why you're still insisting on wasting time and energy on already optimized files? E.g. in my last job there was an optimization made in CI: if _package.json_ is unchanged, then a cache of _node_modules_ is reused (basically just copy the last instance instead of `npm install`). Why wouldn't you apply the same technique? Just pass SVG files that only have changed (usually checked by file time but could be file size too).

Why your workflow depends on Windows? If you can't use grep and other GNU tools (Windows actually have means to utilize them), then why not to write a simple script that handles passing file to SVGO and its output? You can run it on JavaScript just as SVGO runs.

Generally, open source software is based on the rule that every program solves its specific task. The task of SVGO is to optimize SVG images (in terms of file size), not filtering files, text flows, etc. There are special tools for that as it was said before. What you pass to SVGO—that is in the output. No more, no less.

**Comment by SethFalco** [comment]
Hey hey!

I'm inclined to close this as unplanned!

> After considering… my own use cases where _I think_ it can work for me…
> …
> … processes for logs (_I guess_)…
> 
> (emphasis mine)

The writing style seems to me that this is a hypothetical use case where you feel this feature could help in a way that you didn't consider a problem before. And it still doesn't sound like a practical solution over using Git or an external tool to track changes either, or even the JS API as Greli has rightly suggested.

If one doesn't want to use `grep`, they could compare the checksum before and after each SVG as well, which I think would be more stable, or better yet, Git!

> I want to have a workflow run to reduce them

Where is the workflow saving them? I'd imagine in a new PR? If so then ones version control will handle this adequately. Or if it commits directly to `main` then the process is happening out of order, one should review and then push to `main`, not the other way around.

I still don't really understand how this flag would be more ergonomic than other solutions described.

> and regex are generally complex to implement and use and they break easily under changing requirements

Almost anything can break easily under changing requirements! But I respect it, not everyone likes regex! But one can still use checksums, the JS API, Git, etc.

> Generally, open source software is based on the rule that every program solves its specific task

My thoughts towards this are partially inspired by this, and the fact that I'd prefer to keep our features lean so that there's less to maintain and fewer risks of breaking changes. At least in the short to medium-term!
